### PR TITLE
[Ready for testing] Simplified truck parser

### PR DESCRIPTION
**Pull request by only-a-ptr**
Everything seems to be running fine, but bugs always happen.

Known issue: commands are dead. I'm working on it.....

### CODE REVIEW

**Comment by tritonas00** [comment]
http://pastebin.com/ZEwJHwsc

:D

**Comment by tritonas00** [comment]
Well, almost nothing works as it should

Segfaults and many parsing issues :(

**Comment by only-a-ptr** [comment]
:worried:
Weird thing is, everything seems fine on windows. I tried Tatra, GavrilMZ2R, Magnum street.
The only broken thing are commands, that's why doors are loose.
Ulteq also reported segfaults on linux.

**Comment by only-a-ptr** [comment]
@tritonas00 Please try the following:
1. check out 9874b547ba2587baf94824288d953e3144e11965 (Regex-free approach to parsing) and see if the segfaults already happen.
2. if yes, see which of the previous commits did it.
3. if not, look for the rogue commit by halving the interval: check out the middle commit, test it, and then pick another commit closer to the issue.

Thanks.

**Comment by ulteq** [comment]
9874b547ba2587baf94824288d953e3144e11965 works fine for me.

**Comment by tritonas00** [comment]
I pulled this PR, then checked out 9874b54 but i get the `printf_s()` errors. I tried `git cherry pick` to merge only da25705 but i get merge conflicts. Fixing `printf_s()` by hand will take a lot of time to check one by one those commits, since everything before da25705 is not linux friendly.

Any quick way to apply da25705 only  without conflicts? :p

**Comment by ulteq** [comment]
> Any quick way to apply only da25705 without conflicts? :p

I don't think so.

**Comment by only-a-ptr** [comment]
@tritonas00 Sorry about that. I'll rebase everything to purge those `sprintf_s()` from the original commits.

**Comment by tritonas00** [comment]
Ok i tried Desperado Rock Buggy.

75bb102 -> ok
5ed072f -> segfault

Still haunting because Gavril Bandit segfaults with both

**Comment by only-a-ptr** [comment]
BUMP!

I added a very thorough debug logging for the purpose of fixing this PR on Linux.

Additionally, @Michael10055 reported in the Gitter chat that re-downloading all dependencies and creating fresh repo clone produces an error-free build: https://gitter.im/RigsOfRods/rigs-of-rods?at=57e446b5c3e7045a306feab1

Everybody please clean up your environments and re-test.

**Comment by tritonas00** [comment]
http://pastebin.com/4qu6AciE

**Comment by only-a-ptr** [comment]
I just tested all Gavril Bandits in a single session, log: https://www.dropbox.com/s/k60l5en6tvzhr5t/RoR.log?dl=0

The only visible error is ```@@@@@@@@@ SERIOUS @@@@@@@@@@ raw string from truckfile...``` at the start of every parsing, but that's a false alert.

**Comment by tritonas00** [comment]
Here Gavril Bandits and Tatra segfault and the log stops at `17:06:28:   -- PARSER DBG --  >> result: NACA0009.afl`

**Comment by only-a-ptr** [comment]
@tritonas00 I really need to see the whole log, or at least last 50 lines. What section? What line number?

**Comment by tritonas00** [comment]
https://www.dropbox.com/s/nd4iv0vcsov61v1/RoR.log?dl=0

**Comment by only-a-ptr** [comment]
Thanks! 

Very strange thing, line numbers are 1 line off in my log, but correct in your log...

**Comment by tritonas00** [comment]
And with 767 airliner:

https://www.dropbox.com/s/diny4kl3urvtf0s/RoR-air.log?dl=0

Stopped at another .alf this time lol

`17:36:15:   -- PARSER DBG --  >> result: NACA64.1.412.afl`

What is this .alf ?

**Comment by only-a-ptr** [comment]
.AFL -> Airfoil
I'll check the log

**Comment by tritonas00** [comment]
I build with RELEASE. Please try that too.

**Comment by only-a-ptr** [comment]
I built to Release and spawned gavrilbandit200.truck.
No crash, no errors in log. :-/

**Comment by only-a-ptr** [comment]
BUMP!

Any new tests/observations?

I'll try to get myself a linux environment to build and test, but no promises.

**Comment by ulteq** [comment]
This one helped a lot to pinpoint the bug: https://github.com/RigsOfRods/rigs-of-rods/pull/1123/commits/793c83874ff16643c66f689d5944a95ceca52096

```
for (int i = 8; i < 16; i++) { wing.tex_coords[i]   = this->GetArgFloat       (i);  }
```

should have been

```
for (int i = 8; i < 16; i++) { wing.tex_coords[i-8] = this->GetArgFloat       (i);  }
```

-> https://github.com/ulteq/rigs-of-rods/commit/73d69d133d1b876d04a0a1359b33268da64cff97

**Comment by only-a-ptr** [comment]
> ...should have been...

:blush: 

> This one helped a lot to pinpoint the bug: 793c838

Glad it helped

**Comment by tritonas00** [comment]
Great! No segfaults so far here. Only broken door, wheels, steering wheels. Is tatra spawing ok in windows?

**Comment by tritonas00** [comment]
Hmm, some maps segfault like nhelens and flatmap

**Comment by ulteq** [comment]
> Hmm, some maps segfault like nhelens and flatmap

I will take a look at it.

**Comment by only-a-ptr** [comment]
> Does tatra spawn ok in windows?

Yes it does :worried:

**Comment by ulteq** [comment]
nhelens segfaults because of a missing `this->TokenizeCurrentLine();` line here: https://github.com/only-a-ptr/rigs-of-rods/blob/45a7be13fc75b8e2fae42a680c97f6b94fa8a5f7/source/main/resources/rig_def_fileformat/RigDef_Parser.cpp#L1310-L1312

**Comment by ulteq** [comment]
Box5 Rip Saw: screenshot_2016-10-03_15-12-20_1

**Comment by ulteq** [comment]
ParseRotatorsUnified() proposal:

```
void Parser::ParseRotatorsUnified()
{
    if (! this->CheckNumArguments(13)) { return; }

    Rotator2 rotator;
    rotator.inertia_defaults = m_user_default_inertia;
    
    rotator.axis_nodes[0]           = this->GetArgNodeRef( 0);
    rotator.axis_nodes[1]           = this->GetArgNodeRef( 1);
    rotator.base_plate_nodes[0]     = this->GetArgNodeRef( 2);
    rotator.base_plate_nodes[1]     = this->GetArgNodeRef( 3);
    rotator.base_plate_nodes[2]     = this->GetArgNodeRef( 4);
    rotator.base_plate_nodes[3]     = this->GetArgNodeRef( 5);
    rotator.rotating_plate_nodes[0] = this->GetArgNodeRef( 6);
    rotator.rotating_plate_nodes[1] = this->GetArgNodeRef( 7);
    rotator.rotating_plate_nodes[2] = this->GetArgNodeRef( 8);
    rotator.rotating_plate_nodes[3] = this->GetArgNodeRef( 9);
    rotator.rate                    = this->GetArgFloat  (10);
    rotator.spin_left_key           = this->GetArgInt    (11);
    rotator.spin_right_key          = this->GetArgInt    (12);
    
    int offset = 0;

    if (m_current_section == File::SECTION_ROTATORS_2)
    {
        if (! this->CheckNumArguments(16)) { return; }

        if (m_num_args > 13) { rotator.rotating_force  = this->GetArgFloat(13); }
        if (m_num_args > 14) { rotator.tolerance       = this->GetArgFloat(14); }
        if (m_num_args > 15) { rotator.description     = this->GetArgStr  (15); }

        offset = 3;
    }

    this->ParseOptionalInertia(rotator.inertia, 13 + offset);
    if (m_num_args > 17 + offset) { rotator.engine_coupling = this->GetArgFloat(17 + offset); }
    if (m_num_args > 18 + offset) { rotator.needs_engine    = this->GetArgBool (18 + offset); }

    m_current_module->rotators.push_back(rotator);
}
```

**Comment by ulteq** [comment]
ParseCommandsUnified() proposal:

Add `m_current_module->commands_2.push_back(command2);` at the end of the function.

**Comment by only-a-ptr** [comment]
@ulteq Just directly create fixup commits in your clone of this branch please. I'll cherry pick and squash them later. It's also easier to read in diff. If I really dislike something, I can just skip it.

**Comment by ulteq** [comment]
> Just directly create fixup commits in your clone of this branch please.

Agreed.

**Comment by ulteq** [comment]
* `Lotus 76/1` graphics glitch
* `Audi 80 Quattro Group B` exploded on spawn
* `Maitou MLT 731` RoR Bot sitting on the back wheel

**Comment by only-a-ptr** [comment]
I've temporarily just merged Ulteq's fixes and added mine atop of it. Cleanup later.

**Comment by only-a-ptr** [comment]
The last commit is a signifficant fix to deformation threshold of all beams. Everything needs to be re-tested.

That's it for today; I'm off to bed.

File: source/main/resources/rig_def_fileformat/RigDef_Parser.cpp [hunk 1]
```diff
@@ -3098,15 +3098,15 @@ void Parser::ParseProps()
...
     else if (prop.mesh_name.find("dashboard-rh") != std::string::npos) { prop.special = Prop::SPECIAL_DASHBOARD_RIGHT; is_dash = true; }
     else if (prop.mesh_name.find("dashboard"   ) != std::string::npos) { prop.special = Prop::SPECIAL_DASHBOARD_LEFT;  is_dash = true; }
+    else if (Ogre::StringUtil::startsWith(prop.mesh_name, "spinprop") ) { prop.special = Prop::SPECIAL_SPINPROP; }
+    else if (Ogre::StringUtil::startsWith(prop.mesh_name, "pale")     ) { prop.special = Prop::SPECIAL_PALE; }
+    else if (Ogre::StringUtil::startsWith(prop.mesh_name, "seat")     ) { prop.special = Prop::SPECIAL_DRIVER_SEAT; }
+    else if (Ogre::StringUtil::startsWith(prop.mesh_name, "seat2")    ) { prop.special = Prop::SPECIAL_DRIVER_SEAT_2; }
+    else if (Ogre::StringUtil::startsWith(prop.mesh_name, "beacon")   ) { prop.special = Prop::SPECIAL_BEACON; }
+    else if (Ogre::StringUtil::startsWith(prop.mesh_name, "redbeacon")) { prop.special = Prop::SPECIAL_REDBEACON; }
+    else if (Ogre::StringUtil::startsWith(prop.mesh_name, "lightb")   ) { prop.special = Prop::SPECIAL_LIGHTBAR; } // Previously: 'strncmp("lightbar", meshname, 6)'
```
Comment (only-a-ptr) [inline source/main/resources/rig_def_fileformat/RigDef_Parser.cpp]: `lightb` = Nice catch :100:

**Comment by tritonas00** [comment]
So far:

Segfault with 4x4 DAF skiploader

2012 Salf-T-Liner HDX:
https://cloud.githubusercontent.com/assets/2660424/19080727/9e923b66-8a60-11e6-8b27-4f9ba943451a.png

1987 Banshee:
https://cloud.githubusercontent.com/assets/2660424/19080823/fc7d93ec-8a60-11e6-9240-8bb46c356d69.png

Coloring problems and sound glitch with 1990 Chevy k2500 Blazer:
https://cloud.githubusercontent.com/assets/2660424/19081006/7e9d1b5e-8a61-11e6-86fb-b3d2e0647f97.png

**Comment by ulteq** [comment]
Commands description parsing issue:

Truck: MeanMachine Baja 250 LS
Old parser: `description=Shock_Ratio:SOFT<->MED<->HARD`
New parser: `description=Shock_Ratio`

**Comment by only-a-ptr** [comment]
@tritonas00 The "Lowboy" trailer is weird. Current upstream spawns it the same way this new parser does. The old 0.4.0.7 release yields something even worse:
screenshot_1

**Comment by tritonas00** [comment]
About the skiploader segfault i get:

`00:45:07:   -- PARSER DBG --  @@@ ERROR @@@ Overflow in arguments array!! index:2, num_args:2`

http://pastebin.com/k4mtF954

Doesn't happen i current upstream

**Comment by only-a-ptr** [comment]
@tritonas00 I see the bug in code. Fix incoming....

**Comment by only-a-ptr** [comment]
@tritonas00 Fixed. Try now

**Comment by tritonas00** [comment]
Great! I will re-compile tomorrow and check more vehicles :+1:

**Comment by ulteq** [comment]
The skipload still produces those errors:
`@@@ ERROR @@@ Overflow in arguments array!! index:2, num_args:2`

**Comment by only-a-ptr** [comment]
@ulteq  gimme more of that log. What function? What section?

**Comment by ulteq** [comment]
10:18:42:   -- PARSER DBG --  TokenizeCurrentLine() lineNum:835, section:ropes, TEXT {{116,134}}
10:18:42:   -- PARSER DBG --  >> numArgs:2
    [ 0] Start:  0, Len: 3, Text: "116" 
    [ 1] Start:  4, Len: 3, Text: "134" 
10:18:42:   -- PARSER DBG --  GetArgStr() line_num:835, index:1
10:18:42:   -- PARSER DBG --  >> result: 134
10:18:42:   -- PARSER DBG --  GetArgStr() line_num:835, index:2
10:18:42:   -- PARSER DBG --  @@@ ERROR @@@ Overflow in arguments array!! index:2, num_args:2
10:18:42:   -- PARSER DBG --  >> result: ^@^@^@

**Comment by ulteq** [comment]
`John Deere 310 SJ` has multiple issues with hydros, commands, beams ...

**Comment by ulteq** [comment]
**Issues:**

* `John Deere 310 SJ` has multiple issues with hydros, commands, beams ...
* `NEOPLAN tourliner` PARSER DBG --  @@@ ERROR @@@ Overflow in arguments array!!

**Comment by only-a-ptr** [comment]
Closing this as dirty work branch which fulfilled it's purpose. I'll open a new PR with cleaned branch.

**Pull request closed without merging**